balancer.go

```diff
@@ -104,3 +109,44 @@ func (lb *LeastBytes) makeCounters(partitions ...int) (counters []leastBytesCoun
 	})
 	return
 }
+
+var (
+	fnv1aPool = &sync.Pool{
+		New: func() interface{} {
+			return fnv.New32a()
+		},
+	}
+)
+
+// Hash is a Balancer implementation that routes messages to the partition
+// based on the FNV-1a hash of the key.  This ensures that messages with the
+// same key end up on the same partition.
+//
+// Note that this is the same algorithm used by the Sarama Producer and ensures
+// that messages produced by kafka-go will be delivered to the same topics that
+// the Sarama producer would be delivered to
+type Hash struct {
+	rr RoundRobin
+}
+
+func (h *Hash) Balance(msg Message, partitions ...int) (partition int) {
+	if msg.Key == nil {
+		return (&h.rr).Balance(msg, partitions...)
+	}
+
+	hasher := fnv1aPool.Get().(hash.Hash32)
```

Comment:
Sarama allows setting a custom hashing function, I feel like this is almost a requirement because FNV is not a good fit for all keys.

I'd advocate for allowing a hashing function to be configured on the `Hash` structure, something like this:
```go
type Hash struct {
    Func func(key []byte) uint32
}
```
Then a program can set whatever it wants (doing optimizations with a memory pool if needed), and kafka-go defaults Saram's algorithm.

Reply:
Alternatively we could use a field of type `hash.Hash32` since the `Balancer` documentation states that calls to `Balance` are synchronized, we can safely reuse the same hash, calling Reset to reinitialize it.

Reply:
Sure, happy to do it the sarama way.  I saw what sarama had done, but opted not to expose the hasher to minimize the surface area I was touching.

Reply:
Cool, I really think it makes sense here to allow the algorithm to be configurable on the `Hasher` type, it's just one field and doesn't even require making an extra function.